pyro/infer/trace_mean_field_elbo.py

```diff
@@ -0,0 +1,81 @@
+from __future__ import absolute_import, division, print_function
+
+from torch.distributions import kl_divergence
+
+from pyro.distributions.util import is_identically_zero
+from pyro.infer.trace_elbo import Trace_ELBO
+from pyro.infer.util import torch_item
+from pyro.util import warn_if_nan
+
+
+class TraceMeanField_ELBO(Trace_ELBO):
+    """
+    A trace implementation of ELBO-based SVI. This is currently the only
+    ELBO estimator in Pyro that uses analytic KL divergences when those
+    are available.
+
+    In contrast to, e.g.,
+    :class:`~pyro.infer.tracegraph_elbo.TraceGraph_ELBO` and
+    :class:`~pyro.infer.tracegraph_elbo.Trace_ELBO` this estimator places
+    restrictions on the dependency structure of the model and guide.
+    In particular it assumes that the guide has a mean-field structure,
+    i.e. that it factorizes across the different latent variables present
+    in the guide. It also assumes that all of the latent variables in the
+    guide are reparameterized. This latter condition is satisfied for, e.g.,
+    the Normal distribution but is not satisfied for, e.g., the Categorical
+    distribution.
+
+    .. warning:: This estimator may give incorrect results if the mean-field
+      condition is not satisfied.
+
+    Note for advanced users:
+
+    The mean field condition is a sufficient but not necessary condition for
+    this estimator to be correct. The precise condition is that for every
+    latent variable `z` in the guide, its parents in the model must not include
+    any latent variables that are descendants of `z` in the guide. Here
+    'parents in the model' and 'descendants in the guide' is with respect
+    to the corresponding (statistical) dependency structure. For example, this
+    condition is always satisfied if the model and guide have identical
+    dependency structures.
```

Comment:
Would it be sufficient and not too inconvenient to require the model and guide have the same variable ordering, and warn otherwise? This is easy to check from the traces.

Reply:
@fritzo I guess the trace just gives the "coding" order, not "structure" order.

Reply:
@fritzo i think that would be reasonable. probably warn over error/assert